pkg/kube/secretcontroller/secretcontroller.go

```diff
@@ -319,42 +330,51 @@ func (c *Controller) runWorker() {
 }
 
 func (c *Controller) processNextItem() bool {
-	key, quit := c.queue.Get()
+	event, quit := c.queue.Get()
 	if quit {
+		log.Info("secret controller queue is shutting down, so returning")
 		return false
 	}
-	defer c.queue.Done(key)
+	log.Infof("secret controller got event %s from queue for secret %s", event.(secretEvent).event, event.(secretEvent).key)
+	defer c.queue.Done(event)
 
-	err := c.processItem(key.(string))
+	err := c.processItem(event.(secretEvent))
 	if err == nil {
+		log.Debugf("secret controller finished processing event %s for secret %s", event.(secretEvent).event, event.(secretEvent).key)
 		// No error, reset the ratelimit counters
-		c.queue.Forget(key)
-	} else if c.queue.NumRequeues(key) < maxRetries {
-		log.Errorf("Error processing %s (will retry): %v", key, err)
-		c.queue.AddRateLimited(key)
+		c.queue.Forget(event)
+	} else if c.queue.NumRequeues(event) < maxRetries {
+		log.Errorf("Error processing %s (will retry): %v", event, err)
+		c.queue.AddRateLimited(event)
 	} else {
-		log.Errorf("Error processing %s (giving up): %v", key, err)
-		c.queue.Forget(key)
+		log.Errorf("Error processing %s (giving up): %v", event, err)
+		c.queue.Forget(event)
 	}
 
 	return true
 }
 
-func (c *Controller) processItem(key string) error {
-	if key == initialSyncSignal {
+func (c *Controller) processItem(se secretEvent) error {
+	if se.key == initialSyncSignal {
+		log.Info("secret controller initial sync done")
 		c.initialSync.Store(true)
 		return nil
 	}
-
-	obj, exists, err := c.informer.GetIndexer().GetByKey(key)
-	if err != nil {
-		return fmt.Errorf("error fetching object %s error: %v", key, err)
-	}
-
-	if exists {
-		c.addSecret(key, obj.(*corev1.Secret))
+	log.Infof("processing secret event %s for secret %s", se.event.String(), se.key)
+	if se.event == model.EventAdd || se.event == model.EventUpdate {
+		obj, exists, err := c.informer.GetIndexer().GetByKey(se.key)
+		if err != nil {
+			return fmt.Errorf("error fetching object %s error: %v", se.key, err)
+		}
+		// For add and update events, we expect secret in informer cache.
+		// If it is not there it might be the case of it getting deleted and we are processing add/update event now.
+		if !exists {
+			log.Debugf("secret %s does not exist in informer cache, might have been deleted?", se.key)
+		}
+		log.Debugf("secret key %s exists in informer, adding/updating it", se.key)
+		c.addSecret(se.key, obj.(*corev1.Secret))
```

Comment:
If obj = nil, when run into this function it will panic

Reply:
fixed